Userland/Libraries/LibGfx/QOILoader.cpp

```diff
@@ -0,0 +1,285 @@
+    VERIFY((bytes[0] & QOI_MASK_2) == QOI_OP_LUMA);
+    u8 diff_green = (bytes[0] & ~QOI_MASK_2);
+    u8 dr_dg = (bytes[1] & 0b11110000) >> 4;
+    u8 db_dg = (bytes[1] & 0b00001111);
+
+    // Values are stored as unsigned integers with a bias of 32 for the green channel and a bias of 8 for the red and blue channel.
+    return Color {
+        static_cast<u8>(pixel.red() + static_cast<i8>((diff_green - 32) + (dr_dg - 8))),
+        static_cast<u8>(pixel.green() + static_cast<i8>(diff_green - 32)),
+        static_cast<u8>(pixel.blue() + static_cast<i8>((diff_green - 32) + (db_dg - 8))),
+        pixel.alpha(),
+    };
+}
+
+static ErrorOr<u8> decode_qoi_op_run(InputMemoryStream& stream)
+{
+    u8 byte {};
+    stream >> byte;
+    if (stream.handle_any_error())
+        return Error::from_string_literal("QOIImageDecoderPlugin: Invalid QOI image: end of stream while reading QOI_OP_RUN chunk"sv);
+    VERIFY((byte & QOI_MASK_2) == QOI_OP_RUN);
+    u8 run = byte & ~QOI_MASK_2;
+
+    // The run-length is stored with a bias of -1.
+    run += 1;
+
+    // Note that the run-lengths 63 and 64 (b111110 and b111111) are illegal as they are occupied by the QOI_OP_RGB and QOI_OP_RGBA tags.
+    if (run == QOI_OP_RGB || run == QOI_OP_RGBA)
+        return Error::from_string_literal("QOIImageDecoderPlugin: Invalid QOI image: illegal run length"sv);
+
+    VERIFY(run >= 1 && run <= 62);
+    return run;
+}
+
+static ErrorOr<void> decode_qoi_end_marker(InputMemoryStream& stream)
+{
+    u8 bytes[sizeof(END_MARKER)] {};
+    stream >> Bytes { &bytes, sizeof(bytes) };
+    if (stream.handle_any_error())
+        return Error::from_string_literal("QOIImageDecoderPlugin: Invalid QOI image: end of stream while reading end marker"sv);
+    if (!stream.eof())
+        return Error::from_string_literal("QOIImageDecoderPlugin: Invalid QOI image: expected end of stream but more bytes are available"sv);
+    if (memcmp(&END_MARKER, &bytes, sizeof(bytes)) != 0)
+        return Error::from_string_literal("QOIImageDecoderPlugin: Invalid QOI image: incorrect end marker"sv);
+    return {};
+}
+
+static ErrorOr<NonnullRefPtr<Bitmap>> decode_qoi_image(InputMemoryStream& stream, u32 width, u32 height)
+{
+    // FIXME: Why is Gfx::Bitmap's size signed? Makes no sense whatsoever.
+    if (width > NumericLimits<int>::max())
+        return Error::from_string_literal("QOIImageDecoderPlugin: Cannot create bitmap for QOI image of valid size, width exceeds maximum Gfx::Bitmap width"sv);
+    if (height > NumericLimits<int>::max())
+        return Error::from_string_literal("QOIImageDecoderPlugin: Cannot create bitmap for QOI image of valid size, height exceeds maximum Gfx::Bitmap height"sv);
+
+    auto bitmap = TRY(Bitmap::try_create(BitmapFormat::BGRA8888, { width, height }));
+
+    u8 run = 0;
+    Color pixel = { 0, 0, 0, 255 };
+    Color previous_pixels[64] {};
+    for (size_t i = 0; i < 64; ++i)
+        previous_pixels[i] = { 0, 0, 0, 0 };
```

Comment:
The initialisation above should already be setting them to default `Color` instances (probably 0,0,0,255) though I don't see why they need to be initialised here at all?

Reply:
In fact, looks like this is only written to :yakfused:

Reply:
> should already be setting them to default Color instances

yep

> probably 0,0,0,255

That's why I thought I need the loop (`A running array[64] (zero-initialized)`). But the default ctor actually creates 0,0,0,0 so it's fine. 

> though I don't see why they need to be initialised here at all?

Because the spec says it, and because the first chunk could be `QOI_OP_INDEX` - not zero-initializing would lead to undefined decoding behaviour :)

> In fact, looks like this is only written to :yakfused:

Nope:

```cpp
else if ((tag & QOI_MASK_2) == QOI_OP_INDEX)
    pixel = previous_pixels[TRY(decode_qoi_op_index(stream))];
```